[Minlong] iP

seah-minlong opened the pull request:
# Alfred 🤵🏻
> "Some men just want to watch the world burn." - Alfred Pennyworth ([source](https://nolan-batman.fandom.com/wiki/Alfred_Pennyworth/Quote#:~:text=Spoken%20by%20Alfred%20Pennyworth,-%22Because%20some%20men&text=Some%20men%20just%20want%20to%20watch%20the%20world%20burn.%22&text=%22You%20spat%20in%20the%20faces,worse%20before%20they%20got%20better.%22))

Alfred helps you with simple tasks. It's,
- text-based
- easy to learn
- Polite & ~~British~~ Cool

All you need to do is, 
1. download it from [here](https://github.com/seah-minlong/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you. 😉 

And it is **FREE**!


## Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

## Coding

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
``` Java
public static void main(String[] args) {
    new Alfred("./data/Alfred.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

seah-minlong committed:
Replaced Duke with Alfred. Completed Level 0 implementation: Rename, Greet, Exit

seah-minlong committed:
Alfred echos the commands entered by the user, and exits when the user types the command bye

seah-minlong committed:
Add ability to store whatever text entered by the user and display them back to the user when requested.

seah-minlong committed:
Add the ability to mark tasks as done and change status back to not done

- Implemented Task class

seah-minlong committed:
Add support for tracking three types of tasks: ToDos, Deadlines, Events

seah-minlong committed:
Semi-automate testing of Alfred

seah-minlong committed:
Add error handling for incorrect inputs entered by the user

- Add class AlfredException for exceptions specific to Alfred

seah-minlong committed:
Add support for deleting tasks from the list

seah-minlong committed:
Update Alfred such that tasks can now be saved/loaded from the hard disk

seah-minlong committed:
Merge branch 'branch-Level-7'

seah-minlong committed:
Update deadline Task to store deadline as LocalDate instead of String

seah-minlong committed:
Merge branch 'branch-Level-8'

seah-minlong committed:
Update Events to use LocalDate instead of String

seah-minlong committed:
Update Alfred to have Storage and Ui classes

seah-minlong committed:
Update Alfred to have Parser and TaskList classes

seah-minlong committed:
Merge branch 'branch-A-MoreOOP'

seah-minlong committed:
Update Alfred to divde classes into packages

seah-minlong committed:
Merge branch 'branch-A-Packages'

seah-minlong committed:
Merge branch 'add-gradle-support'

seah-minlong committed:
Build and run Alfred using gradle

seah-minlong committed:
Add JUnit tests for 2 different classes

seah-minlong committed:
Add header comments to classes/methods

seah-minlong committed:
Update code to comply with the coding standard

seah-minlong committed:
Enable task search by keyword in description

seah-minlong committed:
Merge branch 'branch-A-JavaDoc'

seah-minlong committed:
Merge branch 'branch-A-CodingStandard'

seah-minlong committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/Alfred/task/Task.java
#	src/main/java/Alfred/ui/Ui.java

Resolution: Adjust comment placement and adding new comments.

seah-minlong committed:
Update Alfred directory to alfred

seah-minlong committed:
Update to include checkstyle and fix style errors

seah-minlong committed:
Merge branch 'branch-A-CheckStyle'

Gilbb commented on src/main/java/alfred/Alfred.java:
```diff
@@ -0,0 +1,116 @@
+                Ui.showFoundTasks(foundTasks);
+                break;
+            default:
+                if (Task.isCreateTaskCommand(input)) {
+                    try {
+                        Task task = Task.initialise(input);
+                        tasks.addTask(task);
+                        Ui.showAddedTaskMessage(task, tasks.getTasksCount());
+                    } catch (AlfredException e) {
+                        Ui.showAlfredError(e);
+                    } catch (Exception e) {
+                        Ui.showUnexpectedError(e);
+                    }
+                } else {
+                    Ui.showInvalidCommand();
+                }
```
Perhaps it would be better to extract this logic into another method to avoid deep nesting? 😄 
```
private void handleDefault() {
   --logic here-
}
```

seah-minlong replied:
I think that is a great suggestion, thanks! 😄

Gilbb commented on src/main/java/alfred/storage/Storage.java:
```diff
@@ -0,0 +1,87 @@
+     * Clears the storage file by deleting it from the file system.
+     * This will remove all saved tasks.
+     */
+    public void clearStorage() {
```
I like this neat addition to clear the storage! 💯

Gilbb commented on src/main/java/alfred/task/Task.java:
```diff
@@ -0,0 +1,157 @@
+     */
+    public static Task initialise(String input) throws AlfredException {
+        String[] parts = input.split(" ");
+        switch (parts[0]) {
+        case "todo":
+            return ToDos.createTask(input);
+        case "event":
+            return Events.createTask(input);
+        case "deadline":
+            return Deadlines.createTask(input);
+        default:
+            throw new AlfredException("Invalid task: " + parts[0]);
+        }
```
Well done to avoid common mistake of having indentation for `case` clauses!

Gilbb reviewed (commented):
Overall, I found that your code is easy to understand and read. Small nitpick regarding the deep nesting, but well done following the Java coding standard!

seah-minlong committed:
Update gui with alfred repeating user input

seah-minlong committed:
Update alfred to work with gui

seah-minlong committed:
Merge branch 'branch-Level-10'

seah-minlong committed:
Update gui with new color scheme

seah-minlong committed:
Merge branch 'branch-Level-10'

seah-minlong committed:
Refactor Task and Deadline to have better code quality

seah-minlong committed:
Refactor code to improve code quality

seah-minlong committed:
Add assertions to validate internal state

The app lacks checks to verify assumptions about program state, like
valid task numbers and non-empty task lists. This can lead to
undetected bugs and invalid operations.

Add assertions to ensure key assumptions about the program state
are enforced, helping catch logical errors during development.

seah-minlong committed:
Merge pull request #1 from seah-minlong/branch-A-Assertions

Add assertions to validate internal state

seah-minlong committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/alfred/Alfred.java

Resolution: Added assertions from master branch into code
from branch-A-CodeQuality.

seah-minlong committed:
Merge pull request #2 from seah-minlong/branch-A-CodeQuality

Improve code quality

seah-minlong committed:
Add workflow

seah-minlong committed:
Merge branch 'branch-A-CI'

seah-minlong committed:
Add ability to add tags

seah-minlong committed:
Add ability to untag Tasks

This commit introduces the functionality to untag tasks in the task
management system. Users can now use the untag command to remove a
specific tag from a task. If the tag is successfully removed,
the updated task is returned.

Similar to the behavior of ArrayList.contains(), when the specified tag
is not found in the task, the task is returned unchanged without
any explicit error message. This decision aligns with the principle
of silent failure when the requested operation doesn't alter the
state, as the absence of the tag is not treated as an error.

seah-minlong committed:
Update task manager to store and load tags

seah-minlong committed:
Merge branch 'C-Tagging'

seah-minlong committed:
Add Factory Classes for Task Creation & rename Task Classes

The code currently mixes task creation logic within the task classes
(Todos, Deadlines, Events), making it harder to maintain and extend
as more task types are introduced. Additionally, the plural class
names are misleading since each class represents a single task.

To improve modularity and clarity, factory classes (TodoFactory,
DeadlineFactory, EventFactory) are introduced to handle task creation
logic separately from the task functionality. Task classes are
renamed to Todo, Deadline, and Event to better reflect that each
represents a single instance.

This refactor simplifies the code by decoupling task creation
from task behaviour, making it easier to manage, extend, and test.
It also improves clarity and consistency by using singular naming
conventions for task classes.

seah-minlong committed:
Merge branch 'branch-TaskFactory'

seah-minlong committed:
Update User Guide and screenshot of Ui

seah-minlong committed:
Update README

seah-minlong committed:
Update Alfred to save tasks after task list changes

Task list was only changing after bye command was given. This was
not in line with the requirements.

Update Alfred to save tasks after every command that changes
the task list. Error message will be displayed if there were errors
saving tasks.

seah-minlong committed:
Merge branch 'branch-fixSaveTasks'
